crates/e2e/src/builders.rs

```diff
@@ -0,0 +1,176 @@
+        FromAccountId,
+    },
+    Environment,
+};
+use scale::Encode;
+
+/// The type returned from `ContractRef` constructors, partially initialized with the execution
+/// input arguments.
+type CreateBuilderPartial<E, Args, R> = CreateBuilder<
+    E,
+    Unset<<E as Environment>::Hash>,
+    Unset<u64>,
+    Unset<<E as Environment>::Balance>,
+    Set<ExecutionInput<Args>>,
+    Unset<ink_env::call::state::Salt>,
+    R,
+>;
+
+/// Shim onto the `CreateBuilder` to allow access to the `ExecutionInput` arguments.
+pub struct ConstructorBuilder<E: Environment, Args: Encode, R> {
+    inner: CreateBuilderPartial<E, Args, R>,
+}
+
+impl<E: Environment, Args: Encode, R> From<CreateBuilderPartial<E, Args, R>>
+    for ConstructorBuilder<E, Args, R>
+{
+    fn from(inner: CreateBuilderPartial<E, Args, R>) -> Self {
+        Self { inner }
+    }
+}
+
+impl<E: Environment, Args: Encode, R> ConstructorBuilder<E, Args, R> {
+    /// Returns encoded constructor arguments.
+    pub fn exec_input(self) -> Vec<u8> {
+        // set all the other properties to default values, we only require the `exec_input`.
+        self.inner
+            .endowment(0u32.into())
+            .code_hash(ink_primitives::Clear::clear())
+            .salt_bytes(Vec::new())
+            .params()
+            .exec_input()
+            .encode()
+    }
+}
+
+#[derive(Debug, Clone)]
+pub struct Message<E: Environment, RetType> {
+    account_id: E::AccountId,
+    exec_input: Vec<u8>,
+    marker: std::marker::PhantomData<RetType>,
+}
+
+impl<E, RetType> Message<E, RetType>
+where
+    E: Environment,
+{
+    pub fn build<F, Ref, Args>(
+        account_id: E::AccountId,
+        message: F,
+    ) -> Message<E, RetType>
+    where
+        E: Environment,
+        Ref: TraitCallBuilder + FromAccountId<E>,
+        Args: scale::Encode,
+        F: FnOnce(
+            &<Ref as TraitCallBuilder>::Builder,
+        ) -> CallBuilder<
+            E,
+            Set<Call<E>>,
+            Set<ExecutionInput<Args>>,
+            Set<ReturnType<RetType>>,
+        >,
+        RetType: scale::Decode,
+    {
+        let contract_ref = <Ref as FromAccountId<E>>::from_account_id(account_id.clone());
+        let call_builder = <Ref as TraitCallBuilder>::call(&contract_ref);
+        let builder = message(&call_builder);
+        let exec_input = builder.params().exec_input().encode();
+        Message {
+            account_id,
+            exec_input,
+            marker: Default::default(),
+        }
+    }
+
+    pub fn account_id(&self) -> &E::AccountId {
+        &self.account_id
+    }
+
+    pub fn exec_input(&self) -> &[u8] {
+        &self.exec_input
+    }
+}
+
+/// Convenience method for building messages for the default environment
+pub fn build_message<Ref>(
+    account_id: <ink_env::DefaultEnvironment as Environment>::AccountId,
+) -> MessageBuilder<ink_env::DefaultEnvironment, Ref>
+where
+    Ref: TraitCallBuilder + FromAccountId<ink_env::DefaultEnvironment>,
```

Comment:
Seems strange to me that everything else is generic over the environment but here we enforce the `DefaultEnvironment`.

I guess most users would end up using this as `build_message<DefaultEnvironment, ContractRef>` anyways, so maybe it's okay

Reply:
Indeed it is a convenience method so the user does not have to specify that generic parameter, which at the moment is hardcoded in the codegen https://github.com/paritytech/ink/blob/e1f10786be6399cafd68899a82635e6968d8075b/crates/e2e/macro/src/codegen.rs#L151.

Trying to make the creating of call messages as ergonomic as possible. Not there yet but this is a start.